### PR TITLE
[RSDK-3540] add sensor reading

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-3540)

1. wire up viam_carto_add_sensor_reading to CartoFacade::AddSensorReading
2. have viam_carto_add_sensor_reading return an error if the sensor reading is invalid
3. have viam_carto_add_sensor_reading detect if the map_builder_mutex lock is held & if so return an error
4. have viam_carto_add_sensor_reading write the sensor reading to the map builder's trajectory build if the reading is valid & the lock is able to be aquired

### CODE REVIEW

File: viam-cartographer/src/carto_facade/io.cc [hunk 1]
```diff
@@ -1,15 +1,17 @@
 // This is an experimental integration of cartographer into RDK.
 #include "io.h"
 
-#include <pcl/io/pcd_io.h>
+#include <pcl/console/time.h>  // pcl::console::TicToc
+#include <pcl/io/pcd_io.h>     // pcl::PCDReader
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/io.cc:5]: remove

File: viam-cartographer/src/carto_facade/io.cc [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
 #include <pcl/point_types.h>
 #include <stdio.h>
 
 #include <boost/filesystem.hpp>
+#include <exception>
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/io.cc:10]: remove

File: viam-cartographer/src/carto_facade/io.cc [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
 #include <fstream>  // std::ifstream
 #include <iomanip>
 #include <iostream>
-#include <string>
+#include <utility>
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/io.cc:14]: remove

File: viam-cartographer/src/carto_facade/io.h [hunk 4]
```diff
@@ -9,6 +9,7 @@
 #include <ostream>
 #include <ratio>
 #include <string>
+#include <tuple>
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/io.h:12]: remove

File: viam-cartographer/src/carto_facade/io.h [hunk 5]
```diff
@@ -23,7 +24,6 @@ static const std::string time_format = "%Y-%m-%dT%H:%M:%S.0000Z";
 // millisecond resolution.
 const std::string MakeFilenameWithTimestamp(std::string path_to_dir,
                                             std::time_t t);
-
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/io.h:26]: remove

File: viam-cartographer/src/carto_facade/util_test.cc [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+BOOST_AUTO_TEST_CASE(carto_sensor_reading_too_few_points_binary_failure) {
+    std::vector<std::vector<double>> too_few_points = {
+        {0.007000, 0.006000, 0.001000, 16711938}};
+
+    auto [success, _] = carto_sensor_reading(help::binary_pcd(too_few_points),
+                                             16409988000001121);
+    BOOST_TEST(!success);
+}
+
+BOOST_AUTO_TEST_CASE(carto_sensor_reading_too_many_points_ascii_failure) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/util_test.cc:105]: rename

File: viam-cartographer/src/carto_facade/util_test.cc [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+    auto [success, timed_pcd_from_string] = carto_sensor_reading(
+        help::ascii_pcd(too_many_points), 16409988000001121);
+    BOOST_TEST(timed_pcd_from_string.ranges.size() == 3);
+    BOOST_TEST(success);
+}
+
+// The lib we use will parse as many points as the header specifies
+// and ignore any others
+BOOST_AUTO_TEST_CASE(carto_sensor_reading_too_many_points_binary_success) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/util_test.cc:120]: krename

File: viam-cartographer/src/carto_facade/util_test.cc [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+#include "test_helpers.h"
+namespace help = viam::carto_facade::test_helpers;
+
+namespace viam {
+namespace carto_facade {
+namespace util {
+
+BOOST_AUTO_TEST_SUITE(CartoFacade_io_demo, *boost::unit_test::disabled())
+
+BOOST_AUTO_TEST_CASE(carto_sensor_reading_binary_success) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/util_test.cc:21]: move below failure tests

File: viam-cartographer/src/carto_facade/util_test.cc [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        {0.007000, 0.006000, 0.001000},
+        {0.007000, 0.006000, 0.001000},
+        {0.007000, 0.006000, 0.001000}};
+
+    auto [success, _] = carto_sensor_reading(
+        help::binary_pcd(wrong_point_shape), 16409988000001121);
+    BOOST_TEST(!success);
+}
+
+BOOST_AUTO_TEST_CASE(TimedPointCloudDataFromPCDBuilder_ascii_success) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/util_test.cc:157]: rename to carto_sensor_reading_ascii_success

File: viam-cartographer/src/carto_facade/util_test.cc [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        carto_sensor_reading(pcd, 16409988000001121);
+    BOOST_TEST(success);
+    BOOST_TEST(timed_pcd_from_string.ranges.size() == points.size());
+    help::timed_pcd_contains(timed_pcd_from_string, points);
+    BOOST_TEST(timed_pcd_from_string.origin == Eigen::Vector3f::Zero());
+    BOOST_TEST(timed_pcd_from_string.time ==
+               cartographer::common::FromUniversal(16409988000001121));
+}
+
+BOOST_AUTO_TEST_CASE(TimedPointCloudDataFromPCDBuilder_binary_ascii_compare) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/util_test.cc:194]: rename to carto_sensor_reading_binary_ascii_compare

File: viam-cartographer/src/carto_facade/carto_facade.h [hunk 11]
```diff
@@ -82,7 +82,7 @@ typedef struct viam_carto_sensor_reading {
     bstring sensor;
     // TODO: change to void* and a size
     bstring sensor_reading;
-    unsigned long long sensor_reading_time_unix_micro;
+    int64_t sensor_reading_time_unix_micro;
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/carto_facade.h:85]: @kim-mishra I decided to change this type to be int64_t given that that 

`time.Now().UnixMicro()` https://pkg.go.dev/time#Duration.Microseconds returns an int64 signed 64 bit integer, and previously the C type was an unsigned 64 bit integer. I also figured that using `int64_t` would be more semantically meaningful than `long long` for those not familiar with C.

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 12]
```diff
@@ -60,7 +60,7 @@ std::ostream &operator<<(std::ostream &os, const ActionMode &action_mode) {
     return os;
 }
 
-std::string std_string_from_bstring(bstring b_str) {
+std::string to_std_string(bstring b_str) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/carto_facade.cc:63]: I figured the shorter name would be easier to follow

File: viam-cartographer/src/carto_facade/util.cc [hunk 13]
```diff
@@ -25,6 +31,116 @@ void write_int_to_buffer_in_bytes(std::string &buffer, int d) {
...
+    try {
+        int err = read_pcd(sensor_reading, blob);
+        if (err) {
+            return {false, point_cloud};
+        }
+    } catch (std::exception &e) {
+        LOG(ERROR) << "exception thrown during read_pcd: " << e.what();
+        return {false, point_cloud};
+    }
+    LOG(INFO) << "read_pcd succeeded";
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/util.cc:119]: Change to verbose

File: viam-cartographer/src/carto_facade/map_builder.h [hunk 14]
```diff
@@ -108,7 +113,7 @@ class MapBuilder {
     double GetRotationWeight();
 
     std::unique_ptr<cartographer::mapping::MapBuilderInterface> map_builder_;
-    cartographer::mapping::TrajectoryBuilderInterface* trajectory_builder;
+    cartographer::mapping::TrajectoryBuilderInterface *trajectory_builder;
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/map_builder.h:116]: lint

File: viam-cartographer/src/carto_facade/map_builder.h [hunk 15]
```diff
@@ -51,11 +51,11 @@ class MapBuilder {
     // SaveMapToStream converted the saved pbstream to a stream and deletes the
     // file.
     std::string ConvertSavedMapToStream(
-        const std::string filename_with_timestamp, std::string* buffer);
+        const std::string filename_with_timestamp, std::string *buffer);
 
     // TryFileClose attempts to close an opened ifstream, returning an error
     // string if it fails.
-    std::string TryFileClose(std::ifstream& file, std::string filename);
+    std::string TryFileClose(std::ifstream &file, std::string filename);
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/map_builder.h:58]: lint

File: viam-cartographer/src/carto_facade/io.cc [hunk 16]
```diff
@@ -95,7 +94,7 @@ double ReadTimeFromTimestamp(std::string timestamp) {
         double sub_sec = 0;
         try {
             sub_sec = (double)std::stof(timestamp.substr(sub_sec_index), &sz);
-        } catch (std::exception& e) {
+        } catch (std::exception &e) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade/io.cc:97]: lint

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 17]
```diff
@@ -627,8 +627,46 @@ int CartoFacade::Stop() {
...
 };
 
-int CartoFacade::AddSensorReading(viam_carto_sensor_reading *sr) {
-    return VIAM_CARTO_SUCCESS;
+void CartoFacade::AddSensorReading(const viam_carto_sensor_reading *sr) {
+    std::string sensor = to_std_string(sr->sensor);
+    if (config.sensors[0] != sensor) {
+        VLOG(1) << "expected sensor: " << sensor << " to be "
+                << config.sensors[0];
+        throw VIAM_CARTO_SENSOR_NOT_IN_SENSOR_LIST;
```
Comment (kim-mishra) [inline viam-cartographer/src/carto_facade/carto_facade.cc:635]: are we expecting only one sensor? also is it possible for this to get called before config validation checks that the length of the sensor list != 0?

File: viam-cartographer/src/carto_facade/carto_facade.cc [hunk 18]
```diff
@@ -627,8 +627,46 @@ int CartoFacade::Stop() {
...
+        }
+        map_builder_mutex.unlock();
+        if (update_latest_global_pose) {
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            latest_global_pose = tmp_global_pose;
+        }
+        return;
+    } else {
+        throw VIAM_CARTO_UNABLE_TO_ACQUIRE_LOCK;
+    }
```
Comment (kim-mishra) [inline viam-cartographer/src/carto_facade/carto_facade.cc:669]: Will this function be called during offline mode?

**Review by kim-mishra — COMMENTED** [review]
Mostly lgtm, jsut a couple of questions

File: viam-cartographer/src/carto_facade/carto_facade.cc (hunk 17 above)
Comment (nicksanford) [reply viam-cartographer/src/carto_facade/carto_facade.cc:635]: 1. For now yes. In the slam_service only the first element of the sensors list is provided by the viam-server json.config will ever be used. This attempts to keep the same behavior while leaving an obvious path to accommodate other sensors (such as IMU) in the future.
2. I don't think so given that the sensor list is added as part of the constructor, which will throw an exception if the sensors list is empty.https://github.com/viamrobotics/viam-cartographer/pull/156/files/4a371f514999eeb22509c413b3c3d975dd727a7a#diff-a0af44f9985e7739a638c01fc02f2c543e0f957e732561ea6bcea0dc6891a97fR158

File: viam-cartographer/src/carto_facade/carto_facade.cc (hunk 18 above)
Comment (nicksanford) [reply viam-cartographer/src/carto_facade/carto_facade.cc:669]: yes

**Review by kim-mishra — APPROVED** [review]
LGTM!

**Pull request merged**